File: Data Structures & Algorithms/trapping-rain-water/submission-0.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int n= height.size();
       vector<int>lmax(n,0);
        vector<int>rmax(n,0);
        
       int ans=0;
        for(int i=1;i<n; i++){
            lmax[i]= max(lmax[i-1], height[i-1]);
        }

        
        for(int i=n-2;i>=0; i--){
            rmax[i]= max(rmax[i+1], height[i+1]);
        }

        for(int idx=0; idx<n; idx++){
           int h= min(lmax[idx], rmax[idx])- height[idx];

           ans+= (h>0)?h: 0;
      }

       return ans;
    }
};

File: Data Structures & Algorithms/eating-bananas/submission-0.java
class Solution {

         public boolean solve(int[] piles, int mid, int h){
                    int time=0;

                    for(int i=0; i<piles.length; i++){
                        time+= Math.ceil((double)piles[i]/ mid);
                    }

                    if(time<=h){
                        return true;
                    }

                    return false;
         }
    public int minEatingSpeed(int[] piles, int h) {
           int low=1;
           int high= piles[0];

           for(int i=1; i<piles.length; i++){
             
              high= Math.max(high, piles[i]);
           }
               int ans =Integer.MAX_VALUE;
           while(low<= high){
               int mid= low + (high- low)/2;

               if(solve(piles, mid, h)){
                     ans= Math.min(ans, mid);
                    high= mid-1;
               }
               else{
                        low= mid+1;
               }
              
           }

           return ans;
    }
}
